RPC call blockchain.scripthash.get_first_use

dagurval opened the pull request:
Implements new RPC call blockchain.scripthash.get_firstuse. This RPC
calls returns the hash of the first block scripthash was found in.

It returns a hash of all the transaction that have scripthash as output
in that block.

Note: Does not return transaction spending scripthash as input. Use
get_history for that.

Output from example script:
```
$ ./blockchain_scripthash_get_firstuse.py 
{'block_hash': '000000000000007f6c0e53f4560e15e1a805470144c00e1d81b78543e69c59c4', 'tx_hash': ['491aa3f7fd1f23882c3437898125259d2a0c044bea0f6aa659b8d7d6ab75632a']}
```

gandrewstone commented:
this looks great!  I'm going to try to work with it now.  

But one important change is to give the client the block height AND the block hash.  Its this something the server can easily access?  3 reasons:

1. I could just index into an array of blocks if I'm keeping the headers in RAM, rather than search.

2. A lot of the electrumX APIs require the height, not the hash:
blockchain.block.header, blockchain.block.headers, blockchain.transaction.get_merkle
(for this same reason, all APIs that include a block hash should also include a block height, in particular: blockchain.transaction.get verbose mode)

3. Another use for the height, which is most important to me, is to see if the block is earlier than my checkpoint (if its later than my checkpoint, and the block hash isn't found, then it must be a different fork, which is useful to know quickly since we can't trust the electrumX server).  That way I'd know that I need to rewind my checkpoint to a certain height, so I will header sync backwards to the specified height.  If the height isn't specified, I would need to possibly header sync backwards to the genesis block to prove server is lying about the block hash or on a different fork.  Sure, the electrumX server could lie and tell me height 1 BUT that warns the client that this is going to take some time, and so it can double check the server's claim by asking another server the same question or asking for the merkle proof of this tx at the provided height.

gandrewstone commented:
I've gotten a regression test for the new API working on RegTest chain.

dagurval commented:
Thanks for the feedback! I added block height to the response.

I also changed the method name from `get_firstuse` to `get_first_use`.

> I've gotten a regression test for the new API working on RegTest chain.

Nice, now I don't have to :D

cculianu commented:
Hey -- as I ranted about here: https://github.com/dagurval/electrum-cash-protocol/pull/6

What is the motivation behind the design for this method and/or what wallet use-cases will it enable or provide?

The method as it's designed right now can't be efficiently implemented on ElectrumX or Fulcrum due to the requirement that *input-only* tx's involving the scripthash be **filtered out**.

So.. this looks like it's going to be an Electrs-only extension.. which is fine.  But .. I'm really very curious here:  I'm a bit baffled what a potentially incomplete list of tx's involving the scripthash on the first block it appeared could be used for?  Why not just return 1 tx? Why a list?

dagurval commented:
ElectrsCash has a minimal and space efficient [database schema](https://github.com/BitcoinUnlimited/ElectrsCash/blob/master/doc/schema.md) and does not record the position a tx is at within a block.

Adding inputs also requires additional database lookup (as opposed to filter out for you).

I suspect adding position of a transaction within block to the database may be worth while, as it also would improve merkle proof generation. The space cost is `u32 * all_txs_in_blockchain`.

If it works for @gandrewstone's use case, I can to change it to return only the first transaction. I'd temporarily assume CTOR for all queries and select first by sorting by transaction ID, but fix later having correct block position for all historical blocks.

cculianu commented:
> Adding inputs also requires additional database lookup (as opposed to filter out for you).

Right and for other implementations -- the requirement that it be *only* outputs creates additional processing burden.  Not every implementation bothers to store whether an event in the history was an input or an output -- especially since all RPC methods don't return that information (they just return whether a scripthash was involved in a tx -- not whether it was involved as input, output, or both).

Ok -- to me this is sounding like it will stay an Electrs-only extension, then... unless that requirement to filter out inputs can be relaxed.

cculianu commented:
> ElectrsCash has a minimal and space efficient [database schema](https://github.com/BitcoinUnlimited/ElectrsCash/blob/master/doc/schema.md) and does not record the position a tx is at within a block.

Riight.. yes yes your implementation is amazing. :P

dagurval commented:
Did you read my response? I'm all for returning only one tx. I don't see a problem with not filtering out inputs either. But must we include them?

> Riight.. yes yes your implementation is amazing. :P

Well, I hope to get there :-). 

The end game is to have ElectrsCash bundled and default enabled with BU and have hundreds, or eventually, hopefully thousands of electrum servers deployed. Keeping it space efficient serves that goal.

dagurval commented:
btw I did not know other servers didn't differentiate between scriphash belonging to input or output, that's interesting

cculianu commented:
> But must we include them?

They don't need to be included.. so long as their existence isn't a problem if they are.

My larger question is what are the tx's used for in the context @gandrewstone imagined? Maybe we can get away with not returning any tx's.. :D

cculianu commented:
> btw I did not know other servers didn't differentiate between scriphash belonging to input or output, that's interesting

Yeah, just an implementation detail to save space.  There is a separate utxo dataset that does store outputs per scripthash but that's ephemeral (outputs can be spent).  

A history is permanent so .. when there was no need to store inputs vs outputs -- you compact that down to "TxHash X involved Scripthash Y" as a history event.... to save space.

gandrewstone commented:
@cculianu You are saying that your implementation might return a spend of this scripthash rather than a deposit?

I'd like to point out that any spend can't be the "first use" unless coins are being created from nothing, so we'd have to change the api semantics and name.

You are right, the main purpose of the API is for derivation path discovery.

The purpose of returning a tx hash is so I can optionally prove you aren't lying by subsequently asking for a merkle proof of this tx in this block.  I can't do that with transactions that spend the scripthash because I can't look at the transaction to verify it actually involves the scripthash. I'd have to subsequently request every tx referred to in the inputs.  Or more likely, I'd loop over all tx hashes you gave me looking for one that uses the scripthash as an output.  

I think that its better for the server to do this work internally than for this loop to be run across the network.

If multiple tx occur in the same block, I'm fine with this API just returning a single arbitrary "deposit" transaction (its unknowable and non-causal which one is the first use), rather than all of them.  Does that help your implementation at all?

cculianu commented:
> @cculianu You are saying that your implementation might return a spend of this scripthash rather than a deposit?
> 
> I'd like to point out that any spend can't be the "first use" unless coins are being created from nothing, so we'd have to change the api semantics and name.

Yes, I know.  However if a lot of stuff happened on that block with that address, it may return all of it, including tx's where it's only a spend.

....

Yeah I read the rest of your comment.  I think this will remain an Electrs-only call as implemented. 

Electron Cash for instance would just need to know that an address had tx's.

It later-on does a merkle proof phase when it asks for a real history. 

So this method call is entirely geared to the thing you wanted to do... got it.

That's all I needed to know, thanks. 👍
